### PR TITLE
Add source/sample components to the instrument view

### LINKED ISSUE #23: Include beamline components in instrument view
How can we display other beamline components in the instrument view, not just pixels?
- monitors
- source
- sample

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #23

Basically `instrument_view` will take a dictionary called `components`, which maps the name of something considered "meta" (such that we can establish its position - sc.vector) to one of a bunch of known shapes. shape names are mapped to drawing functions which take care of scaling and orienting the shapes depending upon the beam direction and overall component distribution (distances)

Code in `ess` will allow us to forward these dictionaries on an instrument by instrument basis. One could imaging having a `odin.instrument_view`, which would internally provide the full `components` dictionary.

### CODE REVIEW

**Comment by nvaytet** [comment]
I think I agree with @SimonHeybrock 's suggestion of moving this into the `ess` repo.

I initially thought that it would be ok to do some of the common things here, because the `source_position` and `sample_position` are already special keys we rely on for unit conversions, but actually this is also going away with the new way to handle conversions via the `transform_coords` method.

We would leave the instrument view as it is in scippneutron, so that one can still display non-ESS instruments, but if you want to load all the components for loki, the special keys should be defined on a per-instrument basis.

File: python/src/scippneutron/instrument_view.py [hunk 1]
```diff
@@ -3,10 +3,129 @@
 # @author Neil Vaytet
 
 import numpy as np
+import pythreejs as p3
+import scipp as sc
 
 
-def instrument_view(scipp_obj=None, positions="position", pixel_size=None, **kwargs):
+def _cube(position, message, sz, color):
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:10]: I find the variable naming a little obscure here.
- I think it is not necessarily a `cube`? the box can have different xyz extents?
- `sz` is not very explicit
- `message` sounds like it has some communication aspect, but in fact it's just the text to be displayed above the component?

File: python/src/scippneutron/instrument_view.py [hunk 2]
```diff
@@ -3,10 +3,129 @@
...
+    geometry = p3.BoxGeometry(width=sz[0],
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:11]: I think `sz[0]` is hard to follow. I would either make a dict `{height: 1, ...}` or just use 3 separate variables?

File: python/src/scippneutron/instrument_view.py [hunk 3]
```diff
@@ -3,10 +3,129 @@
...
+    if fixed_width:
+        return fixed_width
+    else:
+        scaling_factor = 1 / 5.0
+        return sc.norm(det_center - component).value * scaling_factor
+
+
+def _add_beamline(scipp_obj, positions_var, scene):
+    sample_width = None
+    det_center = sc.mean(positions_var)
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:87]: I think you can get this information from the camera, and then you wouldn't need to compute the mean anymore.

File: python/src/scippneutron/instrument_view.py [hunk 4]
```diff
@@ -3,10 +3,129 @@
...
+        source_width = _calculate_component_width(det_center,
+                                                  scipp_obj.meta["source_position"],
+                                                  sample_width)
+        source = scipp_obj.meta["source_position"]
+        beamline_comp_positions.append(sc.norm(source - det_center).value)
+        _add_source(tuple(source.value), scene, sz=([source_width] * 3))
+
+    if not beamline_comp_positions:
+        return None
+    return np.max(np.array(beamline_comp_positions)) * 5.0
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:106]: It's a little strange that `_add_beamline` returns a distance between beamline components.
Maybe you can reset the camera in place here? You may need to pass the camera in anyway, if you want to avoid using `positions_var` as suggested above?

File: python/src/scippneutron/instrument_view.py [hunk 5]
```diff
@@ -3,10 +3,129 @@
...
+
+def _add_beamline(scipp_obj, positions_var, scene):
+    sample_width = None
+    det_center = sc.mean(positions_var)
+    beamline_comp_positions = []
+
+    if "sample_position" in scipp_obj.meta:
+        sample_width = _calculate_component_width(det_center,
+                                                  scipp_obj.meta["sample_position"])
+        sample = scipp_obj.meta["sample_position"]
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:93]:
```suggestion
        sample_position = scipp_obj.meta["sample_position"]
```

File: python/src/scippneutron/instrument_view.py [hunk 6]
```diff
@@ -3,10 +3,129 @@
...
+        beamline_comp_positions.append(sc.norm(sample - det_center).value)
+        _add_sample(tuple(sample.value), scene, sz=([sample_width] * 3))
+    if "source_position" in scipp_obj.meta:
+        source_width = _calculate_component_width(det_center,
+                                                  scipp_obj.meta["source_position"],
+                                                  sample_width)
+        source = scipp_obj.meta["source_position"]
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:100]:
```suggestion
        source_position = scipp_obj.meta["source_position"]
```

**Review by nvaytet — CHANGES_REQUESTED** [review]

File: python/src/scippneutron/instrument_view.py [hunk 7]
```diff
@@ -116,13 +118,14 @@ def _get_camera(plt):
 def instrument_view(scipp_obj=None,
                     positions="position",
                     pixel_size=None,
-                    plot_non_pixels=True,
+                    also_plot=None,
```
Comment (SimonHeybrock) [inline python/src/scippneutron/instrument_view.py:121]: How about `components` or something like that?

Comment (OwenArnold) [reply python/src/scippneutron/instrument_view.py]: image
Not done yet, but this gives a flavour of how things would be done

Comment (SimonHeybrock) [reply python/src/scippneutron/instrument_view.py]: 👍

File: python/src/scippneutron/instrument_view.py (hunk 3 above)
Comment (OwenArnold) [reply python/src/scippneutron/instrument_view.py:133]: I did think of this, but was struggling to figure out how to read that information out of the camera. Camera has lookAt, but there doesn't seem to be a way that I could see to read this back without some complex logic based on the frustrum coordinates.

File: python/src/scippneutron/instrument_view.py [hunk 8]
```diff
@@ -1,12 +1,181 @@
...
+
+
+def _plot_components(scipp_obj, additional, positions_var, scene):
+    det_center = sc.mean(positions_var)
+    furthest_key, furthest_distance = _furthest_component(det_center, scipp_obj,
+                                                          additional)
+    # Some scaling to set width according to distance from detector center
+    scaling_factor = 1 / 10.0
+    width = furthest_distance * scaling_factor
+    shapes = {"cube": _box, "cylinder": _cylinder, "disk": _disk_chopper}
```
Comment (nvaytet) [inline python/src/scippneutron/instrument_view.py:139]: Could we just name these `monitor`, `source`, `sample` `chopper` instead of `cube` and `box`, `cylinder` etc...
i.e. standardize the way each instrument represents a monitor and a sample?
Don't know if we also want to standardize the colors for each component?

**Review by nvaytet — CHANGES_REQUESTED** [review]

File: python/src/scippneutron/instrument_view.py (hunk 8 above)
Comment (OwenArnold) [reply python/src/scippneutron/instrument_view.py:139]: Ok will do. My reasoning was that some instruments might have a preference on how they represented components of each type, i.e use a cube for the sample and a cylinder for the source interchangeably - I think we get more convenience with your approach so lets take that. I think plotting all these non-detector components as grey is sufficient. Maybe we can consider upgrading the colours if we get specific requests about this.

Comment (SimonHeybrock) [reply python/src/scippneutron/instrument_view.py]: Hmm, might want to decouple the shape from the category (monitor, source, ...). Maybe just standardize the color<->category mapping?

Comment (SimonHeybrock) [reply python/src/scippneutron/instrument_view.py]: Consider someone with an instrument with a unique component. We would constantly get new request ("I need analyzer crystals", "I need slits", "I need XYZ", ...). I feel *not* having hard-coded names (or exclusively hard-coded) would be the better option.

Comment (nvaytet) [reply python/src/scippneutron/instrument_view.py:139]: So do you want "not hard-coded" or "exclusively hard-coded"?
I didn't really get what was the conclusion.
I think you would slowly build a library of component representations, slits, etc.. and over time I don't think we will get more and more requests, just a bit to start with.

File: python/src/scippneutron/instrument_view.py [hunk 9]
```diff
@@ -21,14 +190,23 @@ def instrument_view(scipp_obj=None, positions="position", pixel_size=None, **kwa
...
-                pixel_size=pixel_size,
-                **kwargs)
+    plt = plot(scipp_obj,
+               projection="3d",
+               positions=positions,
+               pixel_size=pixel_size,
+               **kwargs)
+
+    # Add additional components from the beamline
+    if components and not isinstance(plot, PlotDict):
```
Comment (SimonHeybrock) [inline python/src/scippneutron/instrument_view.py:208]: Is this meant to read `plt`? `plot` is just the function?

Comment (OwenArnold) [reply python/src/scippneutron/instrument_view.py:208]: Yes, that's a bug.

**Comment by SimonHeybrock** [comment]
Given that this is a new feature, I would assume that we will have to iterate on this. Potential further comments may thus be addressed at a later point, I'll merge so we can try this out in practice.
